File: app/src/main/java/com/hamburgcodingschool/weather/WeatherRepository.java
package com.hamburgcodingschool.weather;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import com.hamburgcodingschool.weather.model.City;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

/**
 * @author Teresa Holfeld, teresa.holfeld@gmail.com
 */
public class WeatherRepository {

    private static final String BASE_URL = "https://api.openweathermap.org";

    private final WeatherService service;
    private final String apiKey;
    private WeatherCallback callback;

    public WeatherRepository(String apiKey) {
        Retrofit retrofit = new Retrofit.Builder()
                .baseUrl(BASE_URL)
                .addConverterFactory(GsonConverterFactory.create())
                .build();
        service = retrofit.create(WeatherService.class);
        this.apiKey = apiKey;
    }

    public void setCallback(WeatherCallback callback) {
        this.callback = callback;
    }

    public void getWeather(String city) {
        Call<City> cityCall = service.getWeather(city, apiKey);
        cityCall.enqueue(new Callback<City>() {
            @Override
            public void onResponse(@NonNull Call<City> call, @NonNull Response<City> response) {
                City city = response.body();
                callback.receiveWeather(city);
            }

            @Override
            public void onFailure(@NonNull Call<City> call, @NonNull Throwable t) {
                callback.receiveWeather(null);
            }
        });
    }

    public interface WeatherCallback {

        void receiveWeather(@Nullable City city);
    }
}

File: app/src/main/java/com/hamburgcodingschool/weather/WeatherRepository.kt
package com.hamburgcodingschool.weather

import com.hamburgcodingschool.weather.model.City

import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory

/**
 * @author Teresa Holfeld, teresa.holfeld@gmail.com
 */
open class WeatherRepository(private val apiKey: String) {

    private val service: WeatherService
    private var callback: WeatherCallback? = null

    init {
        val retrofit = Retrofit.Builder()
                .baseUrl(BASE_URL)
                .addConverterFactory(GsonConverterFactory.create())
                .build()
        service = retrofit.create(WeatherService::class.java)
    }

    fun setCallback(callback: WeatherCallback) {
        this.callback = callback
    }

    open fun getWeather(city: String) {
        val cityCall = service.getWeather(city, apiKey)
        cityCall.enqueue(object : Callback<City> {
            override fun onResponse(call: Call<City>, response: Response<City>) {
                val city = response.body()
                callback!!.receiveWeather(city)
            }

            override fun onFailure(call: Call<City>, t: Throwable) {
                callback!!.receiveWeather(null)
            }
        })
    }

    interface WeatherCallback {

        fun receiveWeather(city: City?)
    }

    companion object {

        private val BASE_URL = "https://api.openweathermap.org"
    }
}
